### ISSUE TITLE
Perspective plugin cannot differentiate w.r.t. custom variable affecting positions [❔ other question]

### ISSUE THREAD

**Issue by realWDC** [mitsuba-renderer/mitsuba2#229]
## Summary

I tried to auto-diff with respect to custom variable `PerspectiveCamera::m_my_variable` but encountering a run-time error.

## System configuration

- Platform: Ubuntu 18.04
- Compiler:clang-9
- Python version: 3.7.6
- Mitsuba 2 version: v2.2.1 (commit @ [b196695](https://github.com/mitsuba-renderer/mitsuba2/commit/b196695023897ff7d3e3d9ad2944589524031b5f))
- Compiled variants:
    * `scalar_rgb`
    * `gpu_autodiff_rgb`
- CUDA version: 10.2
- OptiX version: 7.0

## Description

I modified [`perspective.cpp`](https://github.com/mitsuba-renderer/mitsuba2/blob/master/src/sensors/perspective.cpp) by adding a new `Float`-type variable `m_my_variable`. Only four lines were changed:
```cpp
// ...
template <typename Float, typename Spectrum>
class PerspectiveCamera final : public ProjectiveCamera<Float, Spectrum> {
public:
    // ...
    std::pair<Ray3f, Spectrum> sample_ray(...) const override {
        // ...
        ray.o = trafo.translation() * m_my_variable; // Line 163
        // ...
    }
    // ...
    std::pair<RayDifferential3f, Spectrum> sample_ray_differential(...) const override {
        // ...
        ray.o = trafo.translation() * m_my_variable; // Line 191
        // ...
    }
    // ...
    void traverse(TraversalCallback *callback) override {
        Base::traverse(callback);
        // TODO x_fov
        callback->put_parameter("my_variable", m_my_variable); // Line 272
    }

private:
    Float m_my_variable = 1.f; // Line 300, my new variable
    // ...
}; // end of class
```

## Steps to reproduce

1. Modify [`perspective.cpp`](https://github.com/mitsuba-renderer/mitsuba2/blob/master/src/sensors/perspective.cpp) as before; Compile.
2. Modify [`forward_diff.py`](https://github.com/mitsuba-renderer/mitsuba2/blob/master/docs/examples/10_inverse_rendering/forward_diff.py) from example 10 as following:
```python3
# ...
param_0 = params['PerspectiveCamera.my_variable'] # Line 21
# ...
ek.set_gradient(param_0, 1, backward=False) # Line 25
# ...
```
3. Running `python forward_diff.py` yields a run-time error:
```
Traceback (most recent call last):
  File "forward_diff.py", line 36, in <module>
    Float.forward()
RuntimeError: Internal error in Gather::forward()!
terminating with uncaught exception of type std::runtime_error: CUDABackend: referenced unknown variable 132522
Aborted (core dumped)
```

## My efforts and observations

- Reducing `spp` not helping at all, e.g. `image = render(scene, spp=1)`.

- If changing `m_my_varaible` to be a trivial multiplication on `wav_weight`, e.g.
    ```cpp
    return std::make_pair(ray, m_my_variable * wav_weight); // return of function `sample_ray`
    ```
    then auto-diff works. So probably in my case `SurfaceInteraction` is not differentiable any more.

- I tried modifying `scene.cpp` and `scene_optix.inl`, but the error is still there, even after following changes:
    ```cpp
    // scene.cpp: force computing shape gradients
    m_shapes_grad_enabled = true;
    ```
    ```cpp
    // scene_optix.inl: comment out the following three lines (Line 434, 435, 438)
    if constexpr (is_diff_array_v<Float>) {
    // Differentiable SurfaceInteraction needs to be computed outside of the OptiX kernel
    // if (!has_flag(flags, HitComputeFlags::NonDifferentiable) &&
    //     (requires_gradient(ray_.o) || shapes_grad_enabled())) {
        auto pi = ray_intersect_preliminary_gpu(ray_, active);
        return pi.compute_surface_interaction(ray_, flags, active);
    // }
    }
    ```

I must miss something trivial. Could someone offer some insights please?

**Comment by merlinND** [comment]
Hi @realWDC,

If this is now resolved, would you mind explaining what was the issue?

**Comment by realWDC** [comment]
Hi @merlinND and other Mitsuba guys,

Unfortunately this issue has not been resolved. Yet this issue is a simplfied version of my real case, which has been resolved. Thus I closed this issue.

To articulate, there is a relevant "bug": the autodiff rendering function in Python `_render_helper()` ignores aperture samples and is different from its C++ version `Integrator::render_sample()`:

https://github.com/mitsuba-renderer/mitsuba2/blob/4856a804b58264d9dd53d93ce7f9ee1f8109a799/src/python/python/autodiff.py#L30-L40

Notice `sample3=0`. Changing it to `samples3=sampler.next_2d()` resolved my real case, but not the issue posted here.

Considering the unreality of this toy issue, I have decided to move on and sorry for any mess here.

Thanks for reading,
WDC
